keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -1,959 +1,1429 @@
+To upgrade, first uninstall the v1alpha1 COSI Controller as well as any v1alpha1 Drivers.
+Then, deploy the v1alpha2 COSI Controller and desired v1alpha2 Drivers.
 
-1. App portability is still a concern but we believe that deterministic, unique `Bucket` and `BucketAccess` names can be generated and referenced in BRs and BARs.
+Any COSI resources created using a v1alpha1 system will be incompatible as well.
+The COSI project will document the static provisioning workflow.
+Backend buckets previously created using a COSI v1alpha1 system be made accessible using this workflow.
+The COSI project will document how to manually migrate from v1alpha1 to v1alpha2.
 
+### Version Skew Strategy
 
-### Upgrade / Downgrade Strategy
+COSI is out-of-tree, so version skew strategy is N/A
 
-No changes are required on upgrade to maintain previous behaviour.
+## Alternatives Considered
 
-### Version Skew Strategy
+This KEP has had a long journey and many revisions.
+Here we capture the main alternatives and the reasons why we decided on a different solution.
+Any alternative can be reconsidered in the future as needed/desired.
 
-COSI is out-of-tree, so version skew strategy is N/A
+### Automatically mount buckets to Pods
+
+Early iterations of the COSI v1alpha1 spec tried to design a system that could mount bucket information to user Pods automatically, to serve as an analog to Pod PVC mounts.
+Each Pod application may have different means of and needs for connecting to object storage buckets, so there is not a clear way for this to be implemented in a way that could accommodate all drivers and all users.
+
+### Encode BucketAccess connection information in a JSON blob
+
+In COSI v1alpha1, BucketAccess connection/authentication information was encoded in a JSON blob in the BucketAccess's chosen Secret.
+This was intended to be mounted to user Pods as a file, which is considered generally safer than mounting via environment variables.
+However, several driver implementers gave feedback that they were unable to process the JSON blob file into forms suitable for their applications.
+They instead requested that each entry in the JSON blob be encoded as individual Secret data fields that could be loaded as files or environment variables based on the needs of their individual applications.
+COSI v1alpha2 deprecated the JSON blob file in favor of the more flexible approach of encoding each field in a separate Secret data key.
+Both forms are not supported in order to keep development and usage consistent.
+
+### Cross-resource protection finalizers
+
+In the v1alpha2 design cycle, COSI received feedback from sig-storage regarding the finalizers used for PV/PVC protection (`.../pv-protection` and `.../pvc-protection`).
```

Comment:
Is this caution more about, make sure that you can detect orphaned resources rather than people removing finalizers?

What happens in these cases with the PVs? Does volumes remain mounted when they otherwise shouldn't?

Reply:
Jan noted that users often blow the cross-resource finalizers away on PV/PVC resources which causes troubles for the controllers. I'm not intimately familiar with the results of some of these corner cases.

We theorize that users be less likely to blow away annotations, hopefully resulting in a system that is less prone to known abuses. In the larger sense, I suppose orphans are the primary concern, but this also applies to orphaned object store buckets and users. Orphaned k8s resources are easily handled by COSI, but orphaned backend resources can only be handled by drivers or object store admins. We hope that adding protections that aren't as easily defeated by known abuse methods will cut down on this and ultimately help the storage admin.

Reply:
> We theorize that users be less likely to blow away annotations

Sure, but the resources still disappear?

I don't think annotations as a proxy to finalizers really buys you anything. If this is just "we rely on other resource, oh I can see it's gone, I should error out now" then I guess that's ok? But you should assume users will abuse your annotations just as much as the finalizers